bson/src/main/org/bson/codecs/pojo/PojoBuilderHelper.java

```diff
@@ -53,88 +44,105 @@
+            TypeData.Builder paramBuilder = TypeData.builder((Class) pType.getRawType());
+            for (Type argType : pType.getActualTypeArguments()) {
+                getNestedTypeData(paramBuilder, argType);
+            }
+            builder.addTypeParameter(paramBuilder.build());
+        } else if (type instanceof TypeVariable) {
+            builder.addTypeParameter(TypeData.builder(Object.class).build());
+        } else if (type instanceof Class) {
+            builder.addTypeParameter(TypeData.builder((Class) type).build());
+        }
+    }
+
+    private static <T> FieldModelBuilder<T> getFieldBuilder(final Field field, final Class<T> clazz) {
+        return FieldModel.<T>builder(field);
+    }
+
+
     @SuppressWarnings("unchecked")
     static <T> FieldModelBuilder<T> configureFieldModelBuilder(final FieldModelBuilder<T> builder, final Field field) {
         return builder
                 .fieldName(field.getName())
                 .documentFieldName(field.getName())
-                .typeData((TypeData<T>) getFieldTypeDataFromClass(field.getType()))
+                .typeData((TypeData<T>) getTypeData(field.getGenericType(), field.getType()))
                 .annotations(asList(field.getDeclaredAnnotations()))
                 .fieldSerialization(new FieldModelSerializationImpl<T>())
                 .fieldAccessor(new FieldAccessorImpl<T>(field, field.getName()));
     }
 
-    private static <T> FieldModelBuilder<T> getFieldBuilder(final Field field, final Class<T> clazz, final ResolvedType resolvedType) {
-        return FieldModel.<T>builder(field).typeData(getFieldTypeData(TypeData.builder(clazz), resolvedType));
-    }
-
-    private static <T> TypeData<T> getFieldTypeData(final TypeData.Builder<T> builder, final ResolvedType resolvedType) {
-        TypeBindings bindings = resolvedType.getTypeBindings();
-        for (int i = 0; i < bindings.getTypeParameters().size(); i++) {
-            ResolvedType boundType = bindings.getBoundType(i);
-            builder.addTypeParameter(getFieldTypeData(TypeData.builder(boundType.getErasedType()), boundType));
-        }
-        return builder.build();
-    }
-
-    private static TypeParameterMap getTypeParameterMap(final List<String> genericTypeNames, final ResolvedField resolvedField) {
-        int classParamIndex = genericTypeNames.indexOf(resolvedField.getRawMember().getGenericType().toString());
+    private static TypeParameterMap getTypeParameterMap(final List<String> genericTypeNames, final Field field) {
+        int classParamIndex = genericTypeNames.indexOf(field.getGenericType().toString());
         TypeParameterMap.Builder builder = TypeParameterMap.builder();
         if (classParamIndex != -1) {
             builder.addIndex(classParamIndex);
-        } else if (classParamIndex == -1 && !resolvedField.getType().getTypeBindings().getTypeParameters().isEmpty()) {
-            Type type = resolvedField.getRawMember().getGenericType();
+        } else {
```

Comment:
I gather the second condition is redundant with the instanceof check/loop a few lines down?